src/provider/remoteFileSystemProvider.ts

```diff
@@ -359,29 +359,44 @@ class VirtualFileSystem {
         if (fileType && fileType==='doc' && fileEntity) {
             const doc = fileEntity as DocumentEntity;
             const _content = JSON.stringify(new TextDecoder().decode(content));
+            let currentPos = 0;
+            let hisCount = 0;
             if (doc.version && doc.cache) {
                 const update = {
                     doc: doc.name,
                     lastV: doc.lastVersion,
-                    v: doc.version + 1,
+                    v: doc.version,
                     // Reference: services/web/frontend/js/vendor/libs/sharejs.js#L1288
                     hash: (()=>{
                         if (!doc.mtime || Date.now()-doc.mtime>5000) {
                             doc.mtime = Date.now();
+                            const _content = new TextDecoder().decode(content);
                             return require('crypto').createHash('sha1').update(
-                                "blob " + doc.cache.length + "\x00" + doc.cache
+                                "blob " + _content.length + "\x00" + _content
                             ).digest('hex');
                         }
                     })() as string,
+                    
                     op: (()=>{
-                        return Diff.diffChars(doc.cache, _content)
+                        return Diff.diffChars(JSON.parse(doc.cache), JSON.parse(_content))
                                     .map((part) => {
                                         if (part.count) {
-                                            return {
-                                                p: part.count,
-                                                i: part.added ? part.value : undefined,
-                                                d: part.added ? part.value : undefined,
-                                            };
+                                            const _value = part.value;
+                                            let currentCount = part.count;
+                                            currentPos += currentCount;
+                                            if (part.added || part.removed) {
+                                                if (part.removed){
+                                                    hisCount += part.count;
+                                                    currentCount = 0;
+                                                }
+                                                return {
+                                                    p: currentPos - hisCount - currentCount,
+                                                    i: part.added ?  _value  : undefined,
+                                                    d: part.removed ?  _value : undefined,
+                                                };
+                                                
+                                            }
+                                            
```

Comment:
I have tried to make this code snippet more accessible, please help verify my following trial, and update this part correspondingly.

```typescript
(part) => {
    if (part.count) {
        const incCount = part.removed ? 0 : part.count;
        currentPos += incCount;
        if (part.added || part.removed) {
            return {
                p: currentPos - incCount,
                i: part.added ?  part.value  : undefined,
                d: part.removed ?  part.value : undefined,
            };
        }
    }
}
```